app/src/main/java/com/eveningoutpost/dexdrip/insulin/LinearTrapezoidInsulin.java

```diff
@@ -0,0 +1,56 @@
+package com.eveningoutpost.dexdrip.insulin;
+
+import com.google.gson.JsonObject;
+
+import java.util.ArrayList;
+
+public class LinearTrapezoidInsulin extends Insulin {
+    /// curvedata and all timestamps are defined in minutes
+    private long onset;  // when does profile activity starts
+    private long t1;     // when does activity reaches max
+    private long t2;     // when does activity leaves max
+    private long t3;     // when does activity ends
+
+    private double max;
+
+    public LinearTrapezoidInsulin(String n, String dn, ArrayList<String> ppn, String c, JsonObject curveData) {
+        super(n, dn, ppn, c, curveData);
+
+        onset = curveData.get("onset").getAsLong();
+        if (curveData.get("peak").getAsString().contains("-")) {
+            t1 = Integer.parseInt(curveData.get("peak").getAsString().split("-")[0]);
+            t2 = Integer.parseInt(curveData.get("peak").getAsString().split("-")[1]);
+        } else {
+            t1 = Integer.parseInt(curveData.get("peak").getAsString());
+            t2 = t1;
+        }
+        t3 = curveData.get("duration").getAsLong();
+
+        max = 2.0 / (t2 - t1 + t3);
+        maxEffect = t3;
+    }
+
+    public double calculateIOB(double t) {
+        double time = t - onset;
+        if ((0 <= time) && (time < t1))
+            return 1.0 - time * time * max / (2 * t1);
+        else if ((t1 <= time) && (time < t2))
+            return 1.0 + 0.5 * max * t1 - max * time;
+        else if ((t2 <= time) && (time < t3))
+            return 0.5 * (time - t3) * (time - t3) * max / (t3 - t2);
+        else return 0;
+    }
+
+    public double calculateActivity(double t) {
+        double time = t - onset;
+        if ((0 <= time) && (time < t1))
+            return concentration * time * max / t2;
+        else if ((t1 <= time) && (time < t2))
+            return concentration * max;
+        else if ((t2 <= time) && (time < t3))
+            return concentration * (time - t3) * max / (t3 - t2);
```

Comment:
Functional Error: `concentration` should not be a factor here. Insulin doses are in `Unit`s which already take concentration into account. Only if insulin were entered in in `ml` would we need to factor in `concentration`. _Luckily_ all of the current insulin types in the textfile are U100 so this is currently moot, but it will cause big problems as new ones are entered. @gruoner  if you agree, can you please submit a fix.

Reply:
Not anywhere in the code, but I have seen the U200 insulin last week. Plus I distinctly remember doing mental conversions when all I could find was U40 - presumably it is still available there.

Reply:
If you were drawing U40 insulin into a needle, yes you'd need to do calculations to figure out how much volume to draw to get the correct number of units of insulin. But the number of Units you need, afaik, remains constant independent of concentration. e.g. If you need 10U of insulin, you will need 0.1 ml of U100 or 0.05 ml of U200. But in xDrip+, you will just enter "10U" and it doesn't really matter what the concentration of the insulin that you injected was.

Reply:
If you were drawing U40 insulin into a needle, yes you'd need to do calculations to figure out how much volume to draw to get the correct number of units of insulin. But the number of Units you need, afaik, remains constant independent of concentration. e.g. If you need 10U of insulin, you will need 0.1 ml of U100 or 0.05 ml of U200. But in xDrip+, you will just enter "10U" and it doesn't really matter what the concentration of the insulin that you injected was.

Reply:
concentration does really matter because we do a prediction and so both Units and concentration have to be taken into account. There are two ways to solve this issue (and 'cause we don't use insulin products other than U100) i don't know which one seems to be the "correct one".
1. all insulin profiles are "equated" to U100 and doses registered in XDrip are adjusted to the concentration of the insulin product
- pro: just one insulin profile has to be defined in insulin-profiles.txt and not one for each available concentration on the market (worldwide)
- con: user has to multiply really injected dose by multiples of U100 concentration (not really user-friendly with U40 concentrations)
2. for each concentration available a dedicated profile has to be defined in insulin-profiles.txt (might get a very long list)
- pro: user can directly register the dose (s)he injected
- con: we will get a very long list of profiles (all quite similar to some other) and someone has to manage this list

which one is wanted by the community (esp. from those using the nightly build auto-update-stream)?

Reply:
after some thinking of the issue we also can implement a GUI and datamodel extension to allow the user to define concentration right when enabling/disabling un/used profiles.
When registering the treatment the currently defined concentration is also registered and product is used for prediction - its some kind of dynamic concentration definition.....
But then we have to decide which dose (U100-dose or registered dose) we want to report to nightscout or tidepool

Reply:
@stephencmorton - this is precisely what I had to do; away longer than I planned for and that was all that was available. Insulin was available without a script, but to get syringes (marked for U40) one had to jump through hoops, tiny ones, so I used my U100 BDs that I always carry with me "in case".
@gruoner - the biggest problem I had in the above scenario was awareness that if I multiply instead of dividing by 2.5, I will inject 6 times what I need. If I were in that situation again, already annoyed by all the bs thrown my way by local authorities, I'd rather put the concentration in just once and have an unemotional device tell me how much to inject. I also think that activity profile does not depend on concentration, so it could be just a factor. On the other hand, since all xDrip cares about is the _amount_, rather than _volume_, that is really all that needs to be stored. For the RARE occasion _volume_ needs to be recalculated, I think an option ("until further notice assume such'n'such concentration") would suffice, saving on both storage and configuration space.
  I'm a complete newbie with xDrip, **bowing deeply in gratitude**, so above are JUST opinions, not even suggestions!